src/MultiTokenPermit.sol

```diff
@@ -0,0 +1,249 @@
+     * @return timestamp Timestamp when approval was set
+     */
+    function allowance(
+        address owner,
+        address token,
+        address spender,
+        uint256 tokenId
+    ) external view override returns (uint160 amount, uint48 expiration, uint48 timestamp) {
+        bytes32 tokenKey = _getTokenKey(token, tokenId);
+        Allowance memory allowed = allowances[owner][tokenKey][spender];
+        return (allowed.amount, allowed.expiration, allowed.timestamp);
+    }
+
+    /**
+     * @notice Approve a spender for a specific token or collection
+     * @param token Token contract address
+     * @param spender Address to approve
+     * @param tokenId Token ID (0 for ERC20, specific ID for NFT, type(uint256).max for NFT collection-wide approval)
+     * @param amount Amount to approve (ignored for ERC721, used for ERC20/ERC1155)
+     * @param expiration Timestamp when approval expires (0 for no expiration)
+     */
+    function approve(
+        address token,
+        address spender,
+        uint256 tokenId,
+        uint160 amount,
+        uint48 expiration
+    ) external override {
+        bytes32 tokenKey = _getTokenKey(token, tokenId);
+
+        // Use the same validation as PermitBase
+        _validateApproval(msg.sender, tokenKey, token, spender, expiration);
+
+        // Update the allowance
+        allowances[msg.sender][tokenKey][spender] =
+            Allowance({ amount: amount, expiration: expiration, timestamp: uint48(block.timestamp) });
+
+        // Emit standard approval event from IPermit interface - emit original token address
+        emit Approval(msg.sender, token, spender, amount, expiration);
+    }
+
+    /**
+     * @notice Execute approved ERC721 token transfer
+     * @dev Uses a dual-allowance system: first checks for specific token ID approval,
+     *      then falls back to collection-wide approval (set with tokenId = type(uint256).max).
+     *      This allows users to either approve individual NFTs or entire collections.
+     * @param from Token owner address
+     * @param to Transfer recipient address
+     * @param token ERC721 contract address
+     * @param tokenId The unique NFT token ID to transfer
+     */
+    function transferFrom(address from, address to, address token, uint256 tokenId) public override {
+        // Get the encoded identifier for this specific token ID
+        bytes32 encodedId = _getTokenKey(token, tokenId);
+
+        // First, try to update allowance for the specific token ID
+        (, bytes memory revertDataPerId) = _updateAllowance(from, encodedId, msg.sender, 1);
+
+        if (revertDataPerId.length > 0) {
+            // Fallback: if no specific token approval exists, check for collection-wide approval
+            // Collection-wide approval is set by calling approve() with tokenId = type(uint256).max
+            bytes32 collectionKey = bytes32(uint256(uint160(token)));
```

Comment:
`collectionKey` may equal `encodedId` if the `tokenId` is uint256.max. In that case, no need to try `_updateAllowance` anymore.